### ISSUE TITLE
Function/variable signature issues

### ISSUE THREAD

**Issue by tanzanite777** [DEVSENSE/phptools-docs#558]
Type: <b>Bug</b>

A few, relatively minor, issues that nonetheless keep causing me pause:

```
class Bug {
    public true $x;
    public array|false $nope;
    public array $works;

    // 1) Hover gives wrong signature: "function Bug::ref(bool|string $y): bool|string".
    // Correct: "function &Bug::ref(false|string $y): false|string".
    function &ref(string|false $y) : string|false {
        // note: uncomment relevant return statement

        // Type conflict is detected correctly: cannot return array when only string and false is allowed
        //return $this->works;

        // 2) Related error not detected: trying to return true when only string and false is allowed
        //return $this->x;

        // 3) Type conflict not detected: all types are specified explicitly and 'nope' could be 
        // array that return type does not allow. I do not see any reason why not to mark it as error -
        // so developer would know to fix / add a relevant type guard.
        //return $this->nope;
    }
}
```

Extension version: 1.46.15409
VS Code version: Code 1.89.0 (b58957e67ee1e712cebf466b995adf4c5307b2bd, 2024-05-01T02:09:22.859Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz (8 x 3600)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.88GB (4.85GB free)|
|Process Argv|--crash-reporter-id adf79977-8c83-4303-918e-a7dfc5f31bf1|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
tftest:31042121
vscod805cf:30301675
binariesv615:30325510
vsaa593:30376534
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pyreplss1:30897532
pythonmypyd1:30879173
pythoncet0:30885854
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
showvideot:31016892
chatpanelc:31018788
dsvsc021:30996838
bdiig495:31013172
fchga941:31046352
pythonprt:31039817
dwnewjupyter:31039675

```

</details>

**Comment by tanzanite777** [comment]
Yeah, was not sure if tripping on array|bool <-> string|bool would be too strict - i suspect it indeed could be more trouble than help (false positives).

However, having the two explicit subtypes for boolean and return by reference (reminder in case it did not stand out in OP) to be visible in signatures is my primary concern - i keep needing to go sanity-check the actual source code (thous intricacies sometimes do not even get mentioned in the accompanying PHPdoc as signature itself is expected to be sufficient to derive the meaning - that expectation falls short if the signature lies - especially in the return reference case).

Does not happen often, but often enough to finally take the time to report it :D.

**Comment by tanzanite777** [comment]
Actually, there is another function signature issue i have frequently encountered, but was not freshly in my mind when writing OP. For completeness sake - would appreciate if it could also be fixed.

```
class Foo {
    // hover shows wrong signature: "static function Foo::bar(): void"
    static function bar() : never {
        die;
    }
}
```

Rarely causes any real trouble (it is usually fairly obvious from function name that the signature is lying and the actual return "type" is never), but would be nice if fixed.

**Comment by jakubmisek** [comment]
- [x] `never` return type respected in mouse hover
